### PR TITLE
docs: Secure Storage Documentation

### PR DESCRIPTION
**Pull request by jamescfkung**
Signed-off-by: James Kung <james.kung@linaro.org>

### CODE REVIEW

File: documentation/secure_storage.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+in GloblaPlatform’s TEE Internal API specification (here called Trusted
+Storage). This specification mandates that it should be possible to store
+general-purpose data and key material that guarantees confidentiality and
+integrity of the data stored and the atomicity of the operations that modifies
+the storage (atomicity here means that either the entire operation completes
+successfully or no write is done).
+
+## Overview
+
+![Secure Storage System Architecture](images/secure_storage/secure_storage_system_architecture.png
```
Comment (jforissier) [inline documentation/secure_storage.md]: I don't think we should use image files like this. Of course they look neat, but they are difficult to maintain, all the more since the source files for the pictures are not part of the PR. With binary pictures you cannot easily see what's changed from version to version. And IMO they do not bring much more value than simple text drawings.

File: documentation/secure_storage.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+If the compile time flag CFG_ENC_FS is set to 'y', the  data stored in block
+files will be encrypted, otherwise, the data will not be encrypted. The
+information stored in meta file are always encrypted.
+
+## Key Manager
+
+Key manager is an component in TEE filesystem, and is responsible for handling
+data encryption and decryption and also management of the sensitive key
+materials. There are two types of key used by key manager. One is secure storage
+key(SSK) and another one is file encryption key(FEK).
```
Comment (jforissier) [inline documentation/secure_storage.md]: Missing spaces: `key (SSK)`, `key (FEK)`.

File: documentation/secure_storage.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+### Secure Storage Key(SSK)
+
+SSK is a per-device key and is generated and stored in secure memory when OP-TEE
+is booting. SSK is used for protecting FEK, in other words, is used for
+encrypting/decrypting FEK.
+
+SSK is derived by:
+> SSK = HMAC<sub>SHA256</sub> (HUK, Chip ID || "a static string")
+
+The functions to get Hardware unique key(HUK) and chip ID depend on platform
```
Comment (jforissier) [inline documentation/secure_storage.md]: Space

File: documentation/secure_storage.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+implementation.
+
+### File Encryption Key(FEK)
+
+When a new TEE file is created, key manager will generate a new FEK by
+PRNG(pesudo random number generator) for the TEE file and store the encrypted
```
Comment (jforissier) [inline documentation/secure_storage.md]: Space

**Comment by jforissier** [comment]
What about giving the path to the source files somewhere, and briefly explain what they do? (`core/tee/tee_svc_storage.c`, `./core/tee/tee_fs.c` etc.)

File: documentation/secure_storage.md (hunk 1 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: @jforissier How about put both the source files (e.g. .odg format file) and binary files of those images on github? Having the source files can help us at least to easily modify the images.

File: documentation/secure_storage.md (hunk 2 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will fix it

File: documentation/secure_storage.md (hunk 3 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will fix it

File: documentation/secure_storage.md (hunk 4 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will fix it

**Comment by jamescfkung** [comment]
@jforissier giving a briefly explanation for the secure storage source files is a good idea. Will add a section to describe this.

File: documentation/secure_storage.md (hunk 1 above)
Comment (jforissier) [inline documentation/secure_storage.md]: @jamescfkung yes, well, if we have a consensus on keeping the .png, then please add the .odg too.

**Comment by jamescfkung** [comment]
PR updated, rebased and ready for 2nd review

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: documentation/secure_storage.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+- **core/tee/tee_fs_common.c:** TEE file system & REE file operation interface
+- **core/tee/tee_fs_key_manager.c:** Key manager
+- **lib/libutee/:** GlobalPlatform Internal API library
+
+### Basic File Operation Flow
+
+When TA is calling the write function provided by GP Trusted Storage API to
+write data to a persistent object, a corresponding syscall implemented in TEE
+Trusted Storage Service will be called, which in turn will invoke a series of
+TEE file operations to store the data. TEE file system will then encrypt the data
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: will then encrypt the data

==> 

    will then encrypt the data (when CFG_ENC_FS=y)

File: documentation/secure_storage.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+5. The Trusted Storage must provide a minimum level of protection against
+   rollback attacks. It is accepted that the actually physical storage may be in
+   an unsecure areas and so is vulnerable to actions from outside of the TEE.
+   Typically, an implementation may rely on the REE for that purpose (protection
+   level 100) or on hardware assets controlled by the TEE (protection level
+   1000).
+
+### TEE File Structure In Linux File System
+
+![TEE File Structure](images/secure_storage/tee_file_structure.png
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: Last block is block "N-1"

File: documentation/secure_storage.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+"TEE file structure in Linux file system")
+
+OP-TEE by default use "/data/tee/" as the secure storage space in Linux
+file system. For each TA, OP-TEE use the TA's UUID to create a standalone folder
+for it under the secure storage space folder. For a persistent object belonging
+to a specific TA, OP-TEE creates a TEE file folder which's name is object-id
+under the TA folder.
+
+In TEE file folder, there has a meta file and several block files. Meta file is
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: The meta file format should be detailed

File: documentation/secure_storage.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+for storing the information of the TEE file which is used by TEE file system to
+manage the TEE file; block file is for storing the data of the persistent
+object.
+
+If the compile time flag CFG_ENC_FS is set to 'y', the  data stored in block
+files will be encrypted, otherwise, the data will not be encrypted. The
+information stored in meta file are always encrypted.
+
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: Missing:

* why are we using blocks
* default size of the block, which define
* default max number of blocks, which define

File: documentation/secure_storage.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+key (SSK) and another one is file encryption key (FEK).
+
+### Secure Storage Key(SSK)
+
+SSK is a per-device key and is generated and stored in secure memory when OP-TEE
+is booting. SSK is used for protecting FEK, in other words, is used for
+encrypting/decrypting FEK.
+
+SSK is derived by:
+> SSK = HMAC<sub>SHA256</sub> (HUK, Chip ID || "a static string")
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: tell why "a static string"

File: documentation/secure_storage.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+implementation.
+
+### File Encryption Key(FEK)
+
+When a new TEE file is created, key manager will generate a new FEK by
+PRNG (pesudo random number generator) for the TEE file and store the encrypted
+FEK in meta file. FEK is used for encrypting/decrypting the TEE file information
+stored in meta file or the data stored in block file.
+
+### Meta Data Encryption Flow
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: Missing how are generated MetaIV and BlockIV

File: documentation/secure_storage.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+system and delete the old meta file. Any failure occurred in this stage will
+cause the operation failed and no changes will be made.
+
+- **Clean up stage**
+
+In this stage, TEE file system will clean up unnecessary or old block files.
+If any failure occurred, the operation still be treated as success and some
+garbage files may be leaved in Linux file system.
+
+## Reference
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: Missing a section with future works (antirollback, TA access restriction,...)

File: documentation/secure_storage.md (hunk 7 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: @pascal-brand-st-dev do you mean we should put the data structure of meta data here? The meta file format information is already included in "Meta Data Encryption Flow" section.

File: documentation/secure_storage.md (hunk 6 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will fix it

File: documentation/secure_storage.md (hunk 5 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will fix it

File: documentation/secure_storage.md (hunk 8 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will add these information

File: documentation/secure_storage.md (hunk 10 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will add these information

File: documentation/secure_storage.md (hunk 11 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will add the new section

File: documentation/secure_storage.md (hunk 9 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will explain it

File: documentation/secure_storage.md (hunk 7 above)
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: In the ""Meta Data Encryption Flow" section, I only see `Meta Data`, but I cannot see what is the meta data.

**Comment by jamescfkung** [comment]
updated and rebased

File: documentation/secure_storage.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+manage the TEE file; block file is for storing the data of the persistent
+object.
+
+If the compile time flag CFG_ENC_FS is set to 'y', the  data stored in block
+files will be encrypted, otherwise, the data will not be encrypted. The
+information stored in meta file are always encrypted.
+
+The reason why we store the TEE file content in many small blocks is to
+accelerate the file update speed when handling a large file. The block size and
+maximum number of blocks are defined in core/tee/tee_fs_private.h
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: Is it good to provide the name of the defines and the typical size?

Comment (jamescfkung) [inline documentation/secure_storage.md]: sure, will add the information

**Comment by jamescfkung** [comment]
Updated

File: documentation/secure_storage.md [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+
+- **core/tee/tee_svc_storage.c:** TEE trusted storage service calls
+- **core/tee/tee_fs.c:** TEE file operation interface
+- **core/tee/tee_fs_common.c:** TEE file system & REE file operation interface
+- **core/tee/tee_fs_key_manager.c:** Key manager
+- **lib/libutee/:** GlobalPlatform Internal API library
+
+### Basic File Operation Flow
+
+When TA is calling the write function provided by GP Trusted Storage API to
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: When __a__ TA ...

File: documentation/secure_storage.md [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+write data to a persistent object, a corresponding syscall implemented in TEE
+Trusted Storage Service will be called, which in turn will invoke a series of
+TEE file operations to store the data. TEE file system will then encrypt the
+data (when CFG_ENC_FS=y) and send REE file operation commands and the encrypted
+data to TEE supplicant by a series of RPC messages. TEE supplicant will receive
+the messages and help to store the encrypted data accordingly to the Linux file
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: ... and ~~help to~~ store ...

File: documentation/secure_storage.md [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+![TEE File Structure](images/secure_storage/tee_file_structure.png
+"TEE file structure in Linux file system")
+
+OP-TEE by default use "/data/tee/" as the secure storage space in Linux
+file system. For each TA, OP-TEE use the TA's UUID to create a standalone folder
+for it under the secure storage space folder. For a persistent object belonging
+to a specific TA, OP-TEE creates a TEE file folder which's name is object-id
+under the TA folder.
+
+In TEE file folder, there is a meta file and several block files. Meta file is
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: In __a__ TEE file ...

File: documentation/secure_storage.md [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+for storing the information of the TEE file which is used by TEE file system to
+manage the TEE file; block file is for storing the data of the persistent
+object.
+
+If the compile time flag CFG_ENC_FS is set to 'y', the  data stored in block
+files will be encrypted, otherwise, the data will not be encrypted. The
+information stored in meta file are always encrypted.
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: I imagine people will ask: What is the reason for that? I.e, always having the meta data encrypted?

File: documentation/secure_storage.md [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+
+The reason why we store the TEE file content in many small blocks is to
+accelerate the file update speed when handling a large file. The block size
+(FILE_BLOCK_SIZE) and the maximum number of block of a TEE file
+(NUM_BLOCKS_PER_FILE) are defined in core/tee/tee_fs_private.h.
+
+For now, the default block size is 4KB and the maximum number of block of a
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: the second "block" on this line -> __blocks__

File: documentation/secure_storage.md [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+TEE file is 1024.
+
+## Key Manager
+
+Key manager is an component in TEE file system, and is responsible for handling
+data encryption and decryption and also management of the sensitive key
+materials. There are two types of key used by key manager. One is secure storage
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: ... two types of __keys__ ...

File: documentation/secure_storage.md [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+### Secure Storage Key (SSK)
+
+SSK is a per-device key and is generated and stored in secure memory when OP-TEE
+is booting. SSK is used for protecting FEK, in other words, is used for
+encrypting/decrypting FEK.
+
+SSK is derived by:
+> SSK = HMAC<sub>SHA256</sub> (HUK, Chip ID || "static string")
+
+The functions to get Hardware unique key (HUK) and chip ID depend on platform
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: Hardware unique key &rarr; Hardware Unique Key

File: documentation/secure_storage.md [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+    uint32_t backup_version_table[NUM_BLOCKS_PER_FILE / 32];
+};
+```
+
+### Block Data Encryption Flow
+
+![Block Data Encryption](images/secure_storage/block_data_encryption.png
+"Block data encryption")
+
+A new block IV will be generated by PRNG when a block data need to be updated.
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: ... when block data __needs__ to be ...

File: documentation/secure_storage.md [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+The default size of block IV is defined in core/include/tee/tee_fs_key_manager.h
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: [core/include/tee/tee_fs_key_manager.h](https://github.com/OP-TEE/optee_os/blob/master/core/include/tee/tee_fs_key_manager.h)

File: documentation/secure_storage.md [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+
+### Meta Data Encryption Flow
+
+![Meta Data Encryption](images/secure_storage/meta_data_encryption.png
+"Meta data encryption")
+
+A new meta IV will be generated by PRNG when a meta data need to be updated.
+The default size of meta IV is defined in core/include/tee/tee_fs_key_manager.h
+
+The data structure of meta data is defined in core/tee/tee_fs_private.h as
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: [`core/tee/tee_fs_private.h`](https://github.com/OP-TEE/optee_os/blob/master/core/tee/tee_fs_private.h)

File: documentation/secure_storage.md [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+FEK in meta file. FEK is used for encrypting/decrypting the TEE file information
+stored in meta file or the data stored in block file.
+
+### Meta Data Encryption Flow
+
+![Meta Data Encryption](images/secure_storage/meta_data_encryption.png
+"Meta data encryption")
+
+A new meta IV will be generated by PRNG when a meta data need to be updated.
+The default size of meta IV is defined in core/include/tee/tee_fs_key_manager.h
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: [`core/include/tee/tee_fs_key_manager.h`](https://github.com/OP-TEE/optee_os/blob/master/core/include/tee/tee_fs_key_manager.h)

File: documentation/secure_storage.md [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+PRNG (pesudo random number generator) for the TEE file and store the encrypted
+FEK in meta file. FEK is used for encrypting/decrypting the TEE file information
+stored in meta file or the data stored in block file.
+
+### Meta Data Encryption Flow
+
+![Meta Data Encryption](images/secure_storage/meta_data_encryption.png
+"Meta data encryption")
+
+A new meta IV will be generated by PRNG when a meta data need to be updated.
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: need &rarr; __needs__

File: documentation/secure_storage.md [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+
+In this stage, TEE file system will clean up unnecessary or old block files.
+If an error occurs, the operation still be treated as a success but some
+garbage files may be left over in Linux file system.
+
+## Future Work
+
+- **TA storage space isolation**
+
+OP-TEE provide different folders for different TAs in Linux file system for
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: provide &rarr; provides

File: documentation/secure_storage.md [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+storing their own TEE files, but OP-TEE cannot prevent an attacker from
+directly copying a TEE file from one TA's folder to another TA's folder in
+Linux file system. TEE OS should have the ability to detect those kind of
+attack, but for now OP-TEE secure storage did not meet the requirement.
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: did &rarr; doesn't meet this requirement.

File: documentation/secure_storage.md [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+
+A simple solution to detect the attack is using TA's UUID as AAD
+when calculating the tag of meta file, so that OP-TEE can known if a TEE file
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: known &rarr; know

File: documentation/secure_storage.md [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+is belong to a specific TA when the TA try to open the TEE file.
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: is belong &rarr; belongs

File: documentation/secure_storage.md [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+
+- **TEE file renaming attack detection**
+
+OP-TEE create a specific folder under the TA's folder for each TEE file in
```
Comment (jbech-linaro) [inline documentation/secure_storage.md]: create &rarr; creates

File: documentation/secure_storage.md (hunk 16 above)
Comment (jamescfkung) [inline documentation/secure_storage.md]: will add explanation that  the option is for debugging purpose only for TA developers.

**Comment by jamescfkung** [comment]
rebased and updated

File: documentation/secure_storage.md [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+A solution to detect the attack is using TEE filename as AAD when calculating
+the tag of meta file.
+
+- **Rollback attack detection**
+
+An attacker can backup each version of a TEE file directly from Linux file
+system and can replace the TEE file by an old version one sooner or later.
+
+The basic idea of detecting rollback attack is to add write counter both in
+meta file and another storage which has anit-rollback capability such as eMMC
```
Comment (pascal-brand-st-dev) [inline documentation/secure_storage.md]: s/anit-rollback/anti-rollback

**Comment by pascal-brand-st-dev** [comment]
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jamescfkung** [comment]
Fix typo and add "Reviewed-by" tags

**Comment by jbech-linaro** [comment]
@jamescfkung can you please just rebase on master also? After that we will merge this patch.

`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by jamescfkung** [comment]
rebased and ready to merge
